### PR TITLE
Added logic to select always the correct version in Runtime

### PR DESCRIPTION
**Pull request by drodriguezhdez**

### CODE REVIEW

File: src/version-parser.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        const agentVersion = allVersions[i];
+        if(!agentVersion.includes("beta")){
+            return agentVersion;
+        }
+    }
+
+    return ""
+}
+
+async function xml2json(xml:string):Promise<any> {
```
Comment (juan-fernandez) [inline src/version-parser.ts]: this does not need to be async
```suggestion
function xml2json(xml:string):Promise<any> {
```

File: src/version-parser.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    const response = await fetch(url);
+    const json = await xml2json(await response.text());
+
+    const releaseVersion:string = json.metadata.versioning[0].release[0];
+    if(allowBeta || !releaseVersion.includes("beta")){
+        return releaseVersion;
+    }
+
+    const allVersions:string[] = json.metadata.versioning[0].versions[0].version;
+    for (let i = allVersions.length - 1 ; i >= 0; i--) {
```
Comment (juan-fernandez) [inline src/version-parser.ts]: this is probably easier to read:

```javascript
const lastNonBeta = allVersions.slice().reverse().find(version => !version.includes('beta'))
if (!lastNonBeta) {
  return ''
}
return lastNonBeta
```
note that `slice()` is needed if you don't want to mutate the original array

File: src/executor.ts [hunk 3]
```diff
@@ -1,18 +1,23 @@
 import * as exec from '@actions/exec';
 import * as io from '@actions/io';
 import * as tc from '@actions/tool-cache';
+import * as versionParser from './version-parser'
```
Comment (juan-fernandez) [inline src/executor.ts]: usually easier:
```suggestion
import { getVersionToUse } from './version-parser'
```

File: src/executor.ts [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
     const workdir = process.cwd();
-    const scopeAgentPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-agent/"+agentVersion+"/scope-agent-"+agentVersion+".jar");
-    if(!scopeAgentPath.endsWith(".jar")) {
-        await io.mv(scopeAgentPath, scopeAgentPath+".jar");
-    }
 
-    const gradleInstrumentatorPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/0.1.0/scope-instrumentation-for-gradle-0.1.0.jar");
+    const agentVersion = versionParser.getVersionToUse(scopeAgentMetadataURL, allowBeta);
+    const pluginVersion = versionParser.getVersionToUse(scopeGradlePluginMetadataURL, false);
+    const instrVersion = versionParser.getVersionToUse(scopeGradleInstrMetadataURL, false);
```
Comment (juan-fernandez) [inline src/executor.ts]: these return promises but you are not awaiting for them, so I don't think this is going to work:
this should work:
```javascript
const [agentVersion, pluginVersion, instrVersion] = await Promise.all([getVersionToUse(scopeAgentMetadataURL, allowBeta), getVersionToUse(scopeGradlePluginMetadataURL, false), getVersionToUse(scopeGradleInstrMetadataURL, false)])
```

File: src/version-parser.spec.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    expect(await getVersionToUse("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-agent/maven-metadata.xml", false)).not.toBe("");
+});
+
+test('should return gradle instrumentator version', async () => {
+    expect(await getVersionToUse("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/maven-metadata.xml", false)).not.toBe("");
+});
+
+test('should return gradle plugin version', async () => {
+    expect(await getVersionToUse("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-gradle-plugin/maven-metadata.xml", false)).not.toBe("");
+});
```
Comment (juan-fernandez) [inline src/version-parser.spec.ts:13]: ideally you would want to run this in CI, right?

**Review by juan-fernandez — COMMENTED** [review]
left some minor comments. I think there is an important issue though. Is this working?

File: src/version-parser.spec.ts (hunk 5 above)
Comment (drodriguezhdez) [reply src/version-parser.spec.ts:13]: Is there something special to add if you want to run in a CI?

**Comment by drodriguezhdez** [comment]
Yeah, it won't work in GitHub Actions, cause the missed await.
Thanks for the improvements!

File: src/executor.ts [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
-    const gradleInstrumentatorPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/0.1.0/scope-instrumentation-for-gradle-0.1.0.jar");
+    const gradleInstrumentatorPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/"+instrVersion+"/scope-instrumentation-for-gradle-"+instrVersion+".jar");
     if(!gradleInstrumentatorPath.endsWith(".jar")){
         await io.mv(gradleInstrumentatorPath, gradleInstrumentatorPath+".jar");
     }
 
-    await exec.exec("sh -c \"java -jar "+gradleInstrumentatorPath+".jar "+scopeGradlePluginVersion+" "+scopeAgentPath+".jar "+workdir+" \"");
-}
\ No newline at end of file
+    await exec.exec("sh -c \"java -jar "+gradleInstrumentatorPath+".jar "+pluginVersion+" "+agentVersion+" "+workdir+" \"");
```
Comment (juan-fernandez) [inline src/executor.ts]: there used to be two `.jar` but now there is only one. Is this intended?

File: src/executor.ts [hunk 7]
```diff
@@ -1,18 +1,20 @@
...
+export async function instrument(allowBeta:boolean): Promise<void> {
     const workdir = process.cwd();
-    const scopeAgentPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-agent/"+agentVersion+"/scope-agent-"+agentVersion+".jar");
-    if(!scopeAgentPath.endsWith(".jar")) {
-        await io.mv(scopeAgentPath, scopeAgentPath+".jar");
-    }
+    const [agentVersion, pluginVersion, instrVersion] = await Promise.all([getVersionToUse(scopeAgentMetadataURL, allowBeta), getVersionToUse(scopeGradlePluginMetadataURL, false), getVersionToUse(scopeGradleInstrMetadataURL, false)])
 
-    const gradleInstrumentatorPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/0.1.0/scope-instrumentation-for-gradle-0.1.0.jar");
+    const gradleInstrumentatorPath = await tc.downloadTool("https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/"+instrVersion+"/scope-instrumentation-for-gradle-"+instrVersion+".jar");
```
Comment (juan-fernandez) [inline src/executor.ts]: using templates is usually much easier to read:
```suggestion
    const gradleInstrumentatorPath = await tc.downloadTool(`https://repo1.maven.org/maven2/com/undefinedlabs/scope/scope-instrumentation-for-gradle/${instrVersion}/scope-instrumentation-for-gradle-${instrVersion}.jar`);
```

File: src/executor.ts (hunk 6 above)
Comment (juan-fernandez) [inline src/executor.ts]: template literals come to the rescue in these situations:
```suggestion
    await exec.exec(`sh -c "java -jar ${gradleInstrumentatorPath}.jar ${pluginVersion} ${agentVersion} ${workdir} "`);
```

File: src/scope-for-gradle.ts [hunk 8]
```diff
@@ -2,18 +2,15 @@ import * as core from '@actions/core'
...
     try {
-        let dsn = core.getInput("dsn", {required: true})
+        let dsn = core.getInput("dsn", {required: true});
         core.exportVariable("SCOPE_DSN", dsn);
 
         let executeTestPhase = core.getInput("run-tests", {required: true});
         let command = core.getInput("command", {required: true});
 
-        await executor.instrument(SCOPE_AGENT_VERSION, SCOPE_GRADLE_PLUGIN_VERSION);
+        await executor.instrument(false);
```
Comment (juan-fernandez) [inline src/scope-for-gradle.ts:13]: will `allowBeta` ever be `true` ? Or is that some feature we want to add later on?

File: src/scope-for-gradle.ts [hunk 9]
```diff
@@ -2,18 +2,15 @@ import * as core from '@actions/core'
 import * as executor from './executor'
```
Comment (juan-fernandez) [inline src/scope-for-gradle.ts]: this usually makes easier code:
```suggestion
import { instrument } from './executor'
```

File: src/version-parser.spec.ts (hunk 5 above)
Comment (juan-fernandez) [reply src/version-parser.spec.ts:13]: no, nothing, a simple GHA with node setup should work. Something like https://github.com/undefinedlabs/demo-nodejs/blob/master/.github/workflows/test.yml

**Review by juan-fernandez — APPROVED** [review]

File: src/scope-for-gradle.ts (hunk 8 above)
Comment (drodriguezhdez) [reply src/scope-for-gradle.ts:13]: Probably I'll have a branch for beta agents that it won't be public, but maybe we can discuss if we allow use beta release to the user based on a flag in the GHA

File: src/executor.ts (hunk 6 above)
Comment (drodriguezhdez) [reply src/executor.ts]: Yes, it is. The logic has changed

**Pull request merged**